src/background_workers.c

```diff
@@ -120,23 +134,71 @@ void *RedisAI_Run_ThreadMain(void *arg) {
       queueItem **evicted_items = NULL;
       RedisAI_RunInfo **batch_rinfo = NULL;
 
-      // We first evict the front of the queue
+      // We first peek the front of the queue
       queueItem *item = queueFront(run_queue_info->run_queue);
 
-      // TODO DAG BATCHING
-      // If a batch item is a DAG and relies on unrealized outputs, just skip it
-      // and give way to other items, since the client is surely different (given
-      // the fact that commands are blocking for the individual clent, so that
-      // the temporal sequence of commands cannot be broken).
+      // We define a few variables to inform about the next action to take
+      // based on the status of the entries on the queue
+      // Do we need to unblock the client (or clients in case of batching)?
+      int do_unblock = 0;
+      // Do we need to run the entry (or the entries in case of batching)?
+      int do_run = 0;
+      // Is the entry not ready to run due to inputs not being available yet,
+      // so that we need to put the entry back on the queue?
+      int do_retry = 0;
+      // Were all ops in entry executed for the device, so that we can just
+      // remove the entry from the queue and be done with it?
+      int device_complete = 0;
 
       while (item) {
         RedisAI_RunInfo *rinfo = (RedisAI_RunInfo *)item->value;
 
+        if (rinfo->timeout > 0) {
+          int timedOut = __atomic_load_n(rinfo->timedOut, __ATOMIC_RELAXED);
+
+          if (timedOut == 0) {
+            struct timeval now, sub;
+            gettimeofday(&now, NULL); 
+            timersub(&now, &rinfo->queuingTime, &sub);
+            size_t time_msec = sub.tv_sec * 1000 + sub.tv_usec / 1000;
+
+            if (time_msec > rinfo->timeout) {
+              timedOut = 1;
+              __atomic_store_n(rinfo->timedOut, timedOut, __ATOMIC_RELAXED);
+            }
+          }
+
+          if (timedOut == 1) {
+            queueEvict(run_queue_info->run_queue, item);
+
+            int dagRefCount = __atomic_sub_fetch(rinfo->dagRefCount, 1, __ATOMIC_RELAXED);
+            if (dagRefCount == 0 && rinfo->client) {
+              RedisModule_UnblockClient(rinfo->client, rinfo);
+            }
+
+            queueItem *evicted_item = item;
+            item = item->next;
+            RedisModule_Free(evicted_item);
+
+            continue;
+          }
+        }
+
+        // Since we might be looking through the queue for candidates, we need
+        // to reinitialize our findings every time we consider a new item for
+        // running
+        do_unblock = 0;
+        do_run = 0;
+        do_retry = 0;
+        device_complete = 0;
+ 
         // We keep a list of additional items that are evicted from the queue
         // to find oppotunities for batching
         if (evicted_items) {
           array_free(evicted_items);
           array_free(batch_rinfo);
+          *evicted_items = NULL;
+          *batch_rinfo = NULL;
```

Comment:
Right, I actually didn't have it in my arr version. I added the define and restructured the loop so that we don't need to reallocate the arrays at every loop.